sql: add random syntax generator tests

madelynnblue opened the pull request:
This test is designed to be run from a nightly job for some longish time
(~1h). Its goal is to find panics introduced in our implementation. Since
the executor is not the same go routine as the test, the executor has
been changed to include a panic handler that prepends the failed SQL
statement in a panic.

This uses the yacc package from the docs repo that generates the SQL
diagrams. It and the rsg package are included in internal because they
are not needed for the operation of the database, only for a test.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7970)

<!-- Reviewable:end -->

knz commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KOL7CYc9B9UYqyquq1b:1734341587) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
> ```Go
> 			switch item.Value {
> 			case "IDENT":
> 				v = []string{"ident"}
> ```

Let's have a random selection here between (at least) two strings with mixed casing. I want to check casing does not throw things out of the window.

---

*[internal/rsg/yacc/lex.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7_KxxspuP_Ip6CzE:-KOL7_KxxspuP_Ip6CzF:1264340143) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/lex.go#L3)):*
> ```Go
> // Copyright 2011 The Go Authors. All rights reserved.
> // Use of this source code is governed by a BSD-style
> // license that can be found in the LICENSE file.
> ```

Are you copying (part of) the `go yacc` code here? If so I'd suggest adding to the header comment 1) where you found this code and how it could be updated, if needed 2) highlighting that this is not part of CockroachDB, 3) should not be linked in a CockroachDB binary, and 4) the code is merely bundled in the same repo but we're not responsible for it.

---

*[internal/rsg/yacc/node.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8N7lbKPKH7z3Yvry:-KOL8N7lbKPKH7z3Yvrz:-499318459) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/node.go#L3)):*
> ```Go
> // Copyright 2011 The Go Authors. All rights reserved.
> // Use of this source code is governed by a BSD-style
> // license that can be found in the LICENSE file.
> ```

Ditto

---

*[internal/rsg/yacc/parse.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8Phzxhpx_6h9LbJc:-KOL8Phzxhpx_6h9LbJd:-499318459) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/parse.go#L3)):*
> ```Go
> // Copyright 2011 The Go Authors. All rights reserved.
> // Use of this source code is governed by a BSD-style
> // license that can be found in the LICENSE file.
> ```

Ditto

---

*[sql/executor.go, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8Z7wcJyrO1cQThFP:-KOL8Z7wcJyrO1cQThFQ:806008301) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/executor.go#L362)):*
> ```Go
> 	session.planner.semaCtx.Placeholders.Assign(pinfo)
> 
> 	defer func() {
> ```

Please extract this to a separate commit or PR. It's a generally useful addition independently of the RSG.

---

*[sql/rsg_test.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL95fzlOjfjwz5CYUC:-KOL95fzlOjfjwz5CYUD:-945434485) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L70)):*
> ```Go
> 			default:
> 			}
> 			s := r.Generate("stmt", 20)
> ```

Pull the magic value into a named constant.

---

*[sql/rsg_test.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8zvHGywy1LMPzlZS:-KOL8zvI-bMCGrV3ROP6:-1374515319) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L79)):*
> ```Go
> 		}
> 	}
> 	for i := 0; i < 200; i++ {
> ```

Make this configurable via env var, then modify top level makefile to add this env var and a comment to explain what it does.

---

*[sql/rsg_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOLA1-TVVHvMAdFvpDJ:-KOLA1-Uymd5vYpuEWJI:-1807511590) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L107)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if _, err := db.Exec(`CREATE DATABASE IF NOT EXISTS ident; CREATE TABLE IF NOT EXISTS ident.ident ();`); err != nil {
> ```

I think you can add some "ident" column in here.
Actually I'd recommend:
```
create database ident1
create database ident2
create table if not exists ident1.ident1(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
create table if not exists ident2.ident1(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
create table if not exists ident1.ident2(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
create table if not exists ident2.ident2(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
```

Then generate a random selection for identifiers out of `[]string{"ident1","ident2","IdEnT1", "iDeNt2"}`  

Then *also* generate random CREATE TABLE always with 2 columns and also separate CREATE INDEX statements, all using the same identifiers (so they work on the same tables) but with mixed case, and check that we get an error every time (I am 70% suspicious we are missing something with case normalization).

Then also generate some INSERT and DELETE statements for these automatically.

---

*[sql/rsg_test.go, line 121 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9PrK5EVfNYP_VQOz:-KOL9PrLZnxsv1ogG_0i:21974180) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L121)):*
> ```Go
> 			default:
> 			}
> 			targets := r.Generate("target_list", 30)
> ```

ditto re. constant

---

*[sql/rsg_test.go, line 136 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9SEd3K1o5xsEfdOi:-KOL9SEd3K1o5xsEfdOj:-889721054) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L136)):*
> ```Go
> 		}
> 	}
> 	for i := 0; i < 200; i++ {
> ```

dito re. loop bound

---

*[sql/rsg_test.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9Zm3vOnY1ZRcCesR:-KOL9Zm3vOnY1ZRcCesS:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L214)):*
> ```Go
> 		}
> 	}
> 	for i := 0; i < 200; i++ {
> ```

ditto

---

*[sql/rsg_test.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9b5vPRGMZ4ws-dUa:-KOL9b5vPRGMZ4ws-dUb:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L253)):*
> ```Go
> 			default:
> 			}
> 			expr := r.Generate("func_expr_common_subexpr", 30)
> ```

ditto

---

*[sql/rsg_test.go, line 259 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9cPKZI4_cFxkw-Gs:-KOL9cPKZI4_cFxkw-Gt:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L259)):*
> ```Go
> 		}
> 	}
> 	for i := 0; i < 200; i++ {
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

knz commented:
Since #8152 is in this can probably merge soon now.

mjibson commented:
Review status: 5 of 63 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KPRdPENEvEYQeMEmM96:b-31p204) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Let's have a random selection here between (at least) two strings with mixed casing. I want to check casing does not throw things out of the window.
</details>
Similar to my objection to this above, if you are worried about casing causing errors, let's write a test for that. But it doesn't seem like it belongs here.

---

*[internal/rsg/yacc/lex.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7_KxxspuP_Ip6CzE:-KPRdmDJTsfaFN8Oxcdn:b-f6gnn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/lex.go#L3)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Are you copying (part of) the `go yacc` code here? If so I'd suggest adding to the header comment 1) where you found this code and how it could be updated, if needed 2) highlighting that this is not part of CockroachDB, 3) should not be linked in a CockroachDB binary, and 4) the code is merely bundled in the same repo but we're not responsible for it.
</details>
This parser was written by copying the code from the text/template/parse package and heavily modifying it for yacc. That is, I'm using the same framework and setup, but different lex/parse logic. We already have a file like this: https://github.com/cockroachdb/cockroach/blob/develop/util/leaktest/leaktest.go, and that file doesn't have any of the notices you list. I'd prefer to leave this header as is until we have some guidance about a better way to handle this situation.

---

*[sql/executor.go, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8Z7wcJyrO1cQThFP:-KPR0ZjN4sRVDB3RqVZZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/executor.go#L362)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Please extract this to a separate commit or PR. It's a generally useful addition independently of the RSG.
</details>
Done.

---

*[sql/rsg_test.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL95fzlOjfjwz5CYUC:-KPR13Dhbb3NQtEy81vq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L70)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Pull the magic value into a named constant.
</details>
Done.

---

*[sql/rsg_test.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8zvHGywy1LMPzlZS:-KPR1TQ0Q53yaNLZuKMK:bqruo2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L79)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Make this configurable via env var, then modify top level makefile to add this env var and a comment to explain what it does.
</details>
How about a flag since that's where the rest of the stuff that enables RSG is also?

---

*[sql/rsg_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOLA1-TVVHvMAdFvpDJ:-KPRcZUsMr_VnJQhDxVf:bgsmt4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L107)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think you can add some "ident" column in here.
> Actually I'd recommend:
> ```
> create database ident1
> create database ident2
> create table if not exists ident1.ident1(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
> create table if not exists ident2.ident1(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
> create table if not exists ident1.ident2(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
> create table if not exists ident2.ident2(ident1 decimal, ident2 decimal, index ident1(ident1), index ident2(ident2))
> ```
> 
> Then generate a random selection for identifiers out of `[]string{"ident1","ident2","IdEnT1", "iDeNt2"}`  
> 
> Then *also* generate random CREATE TABLE always with 2 columns and also separate CREATE INDEX statements, all using the same identifiers (so they work on the same tables) but with mixed case, and check that we get an error every time (I am 70% suspicious we are missing something with case normalization).
> 
> Then also generate some INSERT and DELETE statements for these automatically.
> 
</details>
These new tests you describe may be helpful, but they aren't random syntax tests like these, so I think they should be somewhere else in another PR. RSG tests are only trying to find panics, not errors.

I've added an ident column to the table because I think that has value as it may be specified in some queries.

---

*[sql/rsg_test.go, line 121 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9PrK5EVfNYP_VQOz:-KPR1rIO9lMhqGZ2rs86:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L121)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> ditto re. constant
</details>
Done.

---

*[sql/rsg_test.go, line 136 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9SEd3K1o5xsEfdOi:-KPR1sB9pODAL9n67LO0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L136)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> dito re. loop bound
</details>
Done.

---

*[sql/rsg_test.go, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9Zm3vOnY1ZRcCesR:-KPR1stGUZNKcmyoUhQz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L214)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> ditto
</details>
Done.

---

*[sql/rsg_test.go, line 253 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9b5vPRGMZ4ws-dUa:-KPR21lWjZP8othharX8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L253)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> ditto
</details>
Done.

---

*[sql/rsg_test.go, line 259 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL9cPKZI4_cFxkw-Gs:-KPR22-ku3YP41arT0M7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L259)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KQ2fyXMGtnymuenArFj:bn9uz26) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Similar to my objection to this above, if you are worried about casing causing errors, let's write a test for that. But it doesn't seem like it belongs here.
</details>
Again it will ensure that the NormalizeName method is called everywhere it should. Of course we can also write tests but due to the sheer number of places where NormalizeName is used we are likely to forget a few tests and that's what fuzz testing can help us detect.

---

*[internal/rsg/yacc/lex.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7_KxxspuP_Ip6CzE:-KQ2gFseVXnKJVfJclNa:b-4jx4mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/lex.go#L3)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> This parser was written by copying the code from the text/template/parse package and heavily modifying it for yacc. That is, I'm using the same framework and setup, but different lex/parse logic. We already have a file like this: https://github.com/cockroachdb/cockroach/blob/develop/util/leaktest/leaktest.go, and that file doesn't have any of the notices you list. I'd prefer to leave this header as is until we have some guidance about a better way to handle this situation.
</details>
I think you can still *append* a comment that states exactly what you explain above "this code was produced by copying the code from .... "

---

*[sql/rsg_test.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8zvHGywy1LMPzlZS:-KQ2fUJYoCj5UYwO0b8j:bmvas5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L79)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> How about a flag since that's where the rest of the stuff that enables RSG is also?
</details>
Yep, perfect.

---

*[sql/rsg_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOLA1-TVVHvMAdFvpDJ:-KQ2f_YtzZaSHcPXmwpm:b-65l68w) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L107)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> These new tests you describe may be helpful, but they aren't random syntax tests like these, so I think they should be somewhere else in another PR. RSG tests are only trying to find panics, not errors.
> 
> I've added an ident column to the table because I think that has value as it may be specified in some queries.
</details>
So what I am proposing is to create the schema described above at the start of the test then let the random generator produce random combinations of ident1 and ident2. This way it will exercise all combinations of name lookups in different parts of expressions (db.table.col, tb.table, table.col etc).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KQ4UaviHU0fS9HkimYW:b-c9tct6) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Again it will ensure that the NormalizeName method is called everywhere it should. Of course we can also write tests but due to the sheer number of places where NormalizeName is used we are likely to forget a few tests and that's what fuzz testing can help us detect.
</details>
What happens if that method was not called where it should be? A panic or an error?

---

*[internal/rsg/yacc/lex.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7_KxxspuP_Ip6CzE:-KQ49rb4Y7k-pjCjC09C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/lex.go#L3)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I think you can still *append* a comment that states exactly what you explain above "this code was produced by copying the code from .... "
</details>
Done.

---

*[internal/rsg/yacc/node.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8N7lbKPKH7z3Yvry:-KQ49pAyoWGL67x806qc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/node.go#L3)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Ditto
</details>
Done.

---

*[internal/rsg/yacc/parse.go, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL8Phzxhpx_6h9LbJc:-KQ49pUQMx25HVO0PKdB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/yacc/parse.go#L3)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Ditto
</details>
Done.

---

*[sql/rsg_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOLA1-TVVHvMAdFvpDJ:-KQ4UTTY2BXF1U0jOlM5:b-bqtd18) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L107)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> So what I am proposing is to create the schema described above at the start of the test then let the random generator produce random combinations of ident1 and ident2. This way it will exercise all combinations of name lookups in different parts of expressions (db.table.col, tb.table, table.col etc).
</details>
If I add this, do you expect panics or errors?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: add random syntax generator tests

This test is designed to be run from a nightly job for some longish time
(~1h). Its goal is to find panics introduced in our implementation. Since
the executor is not the same go routine as the test, the executor has
been changed to include a panic handler that prepends the failed SQL
statement in a panic.

This uses the yacc package from the docs repo that generates the SQL
diagrams. It and the rsg package are included in internal because they
are not needed for the operation of the database, only for a test.

mjibson commented:
I've asked about panic vs error because it will inform where the NormalizeName testing goes. If these things produce panics, then they can go into the existing places. If you expect errors, then they will have to go into a new test that tests for lack of errors, since the current ones can produce valid, error-free statements.

---

Review status: 3 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-:-KQ4VpHbxXjr2E6NUTrF:b-of9kqc)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 3 of 3 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KQ4WJCurDZqxjl6jf4h:b-btj7jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> What happens if that method was not called where it should be? A panic or an error?
</details>
This in combination with the other thing would guarantee that all generated names can be resolved (if there's a failure, it shouldn't be related to name lookup any more). So if we observe a name lookup error after this, there would be a normalization bug or a lookup bug.

---

*[sql/rsg_test.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOLA1-TVVHvMAdFvpDJ:-KQ4WRmIZluH1fHP33Yy:bastzbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/sql/rsg_test.go#L107)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> If I add this, do you expect panics or errors?
</details>
Both actually. (truth be told, I'd rather have neither but oh well). The most important check is that once this schema goes in, there should not be any error related to name look up any more (name ... not found). If we find any, that means a lookup was not implemented properly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[internal/rsg/rsg.go, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970#-KOL7CYc9B9UYqyquq1a:-KQ6a-Ky_OlrQa8YKvY1:b5opn1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/057530a9ecef0270713f94b98b964bff8d5c7f76/internal/rsg/rsg.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This in combination with the other thing would guarantee that all generated names can be resolved (if there's a failure, it shouldn't be related to name lookup any more). So if we observe a name lookup error after this, there would be a normalization bug or a lookup bug.
</details>
Ok. I'll add this stuff in a followup PR. There's enough new stuff here that I'd need to figure out and test that I'd prefer to do it separately since this PR is already like a month old.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7970)*
<!-- Sent from Reviewable.io -->

knz commented:
LGTM! Very nice work.
